msi/msi.go

```diff
@@ -37,23 +47,52 @@ func NewMsiProvider(client httputil.HttpClient) provider {
 	return provider{httpClient: client}
 }
 
-func (p *provider) GetMsi() (Msi, error) {
+func (p *provider) getMsiHelper(queryParams map[string]string) (*Msi, error) {
 	var msi = Msi{}
-	code, body, err := p.httpClient.Get(metadataMsiURL, map[string]string{"Metadata": "true"})
+	requestUrl, err := url.Parse(metadataMsiBaseURL)
+	if err != nil {
+		return &msi, err
+	}
+	urlQuery := requestUrl.Query()
+	for key, value := range queryParams {
+		urlQuery.Add(key, value)
+	}
+	requestUrl.RawQuery = urlQuery.Encode()
+
+	code, body, err := p.httpClient.Get(requestUrl.String(), map[string]string{"Metadata": "true"})
 	if err != nil {
-		return msi, err
+		return &msi, err
 	}
 
 	if code != 200 {
-		return msi, errorhelper.AddStackToError(fmt.Errorf("unable to get msi, metadata service response code %v", code))
+		return &msi, errorhelper.AddStackToError(fmt.Errorf("unable to get msi, metadata service response code %v", code))
 	}
 
 	err = json.Unmarshal(body, &msi)
```

Comment:
under what circumstances will the service return a 200 with an error body? Shouldn't we just rely on the 200, and if we don't receive one then retrieve the body to provide more info on the failure?

Reply:
not that I know of,
err could result form being unable to create the http request, or being unable to open a port. So, I check for err first, then check for status. It gives us a finer level of tracing if anything goes wrong.